{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import tensorqtl\n",
    "from tensorqtl import genotypeio, cis, trans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#----------------------------------------------------------------------\n",
    "#? 基因表达量文件跑peer时，不需要header与index\n",
    "#* 对基因表达量进行进一步过滤\n",
    "#? 表达量大于0.5 的样本占比超过20% \n",
    "#? 存在两倍表达差异\n",
    "#----------------------------------------------------------------------\n",
    "geneExpression=pd.read_csv(\"FPKM/gene_expression.txt\",header=0,index_col=0,sep=\"\\t\")\n",
    "geneExpression=geneExpression.T\n",
    "lowIndex=int(geneExpression.shape[0]*0.05)\n",
    "highIndex=int(geneExpression.shape[0]*0.95)\n",
    "fiterGenePair=[]\n",
    "for gene,ratio in geneExpression.apply(lambda x: len([i for i in x if i>0.1])/geneExpression.shape[0],axis=0).iteritems():\n",
    "        if ratio>=0.05:\n",
    "            lowExpression=geneExpression[gene].sort_values()[lowIndex]\n",
    "            highExpression=geneExpression[gene].sort_values()[highIndex]\n",
    "            #* 基因在至少20%样本中表达\n",
    "            if (highExpression-lowExpression)/(highExpression+lowExpression)>=0.33:\n",
    "                #* 95%和5%的样本存在表达差异\n",
    "                fiterGenePair.append(gene) \n",
    "\n",
    "fiter_expression=geneExpression[fiterGenePair]\n",
    "fiter_expression.to_csv(\"./peer/filter_gene_expression.txt\",header=False,index=False,sep=\",\")\n",
    "pd.Series(fiterGenePair).to_csv(\"peer/filter_genePair.txt\",header=False,index=False,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "using cpu\n"
     ]
    }
   ],
   "source": [
    "#TODO 进行eQTL mapping\n",
    "import pandas as pd\n",
    "import tensorqtl\n",
    "from tensorqtl import genotypeio, cis, trans\n",
    "import os\n",
    "import torch\n",
    "import sys\n",
    "\n",
    "os.environ['CUDA_VISIBLE_DEVICES'] = \"0\"\n",
    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "print('using {}'.format(device))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TODO 输入数据\n",
    "plink_prefix_path ='/cotton/Liuzhenping/diaploidCotton_sQTL/Genotype/filter2_Q600_SNPs_joint_216_number_chr'\n",
    "expression_bed = './phenotype/gene_expression_peer_5.bed.gz'\n",
    "covariates_file = './PCA_qcovar.txt'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Mapping files: 100%|██████████| 3/3 [00:05<00:00,  1.99s/it]\n"
     ]
    }
   ],
   "source": [
    "#TODO 加载数据\n",
    "# load phenotypes and covariates\n",
    "phenotype_df, phenotype_pos_df = tensorqtl.read_phenotype_bed(expression_bed)\n",
    "\n",
    "covariates_df = pd.read_csv(\n",
    "    covariates_file, sep='\\t', index_col=0, header=None)  # samples x covariates\n",
    "# * 协变量文件，最终格式为 samples x covariates\n",
    "covariates_df.drop([1], axis=1)\n",
    "covariates_df = covariates_df.drop([1], axis=1)\n",
    "covariates_df.index.name = 'id'\n",
    "covariates_df.columns = ['PC1', 'PC2', 'PC3']\n",
    "covariates_df = covariates_df.loc[phenotype_df.columns]\n",
    "\n",
    "#* 基因型文件, 对应时期的样本编号\n",
    "pr = genotypeio.PlinkReader(\n",
    "    plink_prefix_path, select_samples=phenotype_df.columns)\n",
    "\n",
    "# load genotypes and variants into data frames\n",
    "genotype_df = pr.load_genotypes()\n",
    "variant_df = pr.bim.set_index('snp')[['chrom', 'pos']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 进行cis-eQTL permutate分析\n",
    "cis_df = cis.map_cis(genotype_df,\n",
    "                         variant_df,\n",
    "                         phenotype_df,\n",
    "                         phenotype_pos_df,\n",
    "                         covariates_df,\n",
    "                         nperm=1000,\n",
    "                         maf_threshold=0.05,\n",
    "                         window=1000000,\n",
    "                         seed=2022\n",
    "                         )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 进行cis-eQTL nominal分析\n",
    "cis.map_nominal(genotype_df, variant_df,\n",
    "                             phenotype_df,\n",
    "                             phenotype_pos_df,\n",
    "                             prefix='nominalOut',\n",
    "                             covariates_df=covariates_df,\n",
    "                             maf_threshold=0.05,\n",
    "                             window=1000000,\n",
    "                             output_dir='.',\n",
    "                             write_top=True, write_stats=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Computing q-values\n",
      "  * Number of phenotypes tested: 7\n",
      "  * Correlation between Beta-approximated and empirical p-values: 0.9995\n",
      "  * Calculating q-values with lambda = 0.010\n",
      "  * Proportion of significant phenotypes (1-pi0): 0.00\n",
      "  * QTL phenotypes @ FDR 0.05: 0\n"
     ]
    }
   ],
   "source": [
    "#TODO 矫正pvalue \n",
    "cis_df=pd.read_csv(\"./All_eGene_permutate.txt\",header=0,index_col=0,sep=\"\\t\")\n",
    "#* 测试的基因数目太少了，修改lambda\n",
    "tensorqtl.calculate_qvalues(cis_df, qvalue_lambda=0.01)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>num_var</th>\n",
       "      <th>beta_shape1</th>\n",
       "      <th>beta_shape2</th>\n",
       "      <th>true_df</th>\n",
       "      <th>pval_true_df</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>tss_distance</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>af</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>pval_perm</th>\n",
       "      <th>pval_beta</th>\n",
       "      <th>qval</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phenotype_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Garb_01G000010</th>\n",
       "      <td>2910</td>\n",
       "      <td>1.060317</td>\n",
       "      <td>167.80818</td>\n",
       "      <td>154.71806</td>\n",
       "      <td>0.002920</td>\n",
       "      <td>1_360122</td>\n",
       "      <td>347814</td>\n",
       "      <td>63</td>\n",
       "      <td>117</td>\n",
       "      <td>0.297927</td>\n",
       "      <td>0.000912</td>\n",
       "      <td>0.271290</td>\n",
       "      <td>0.080528</td>\n",
       "      <td>0.373626</td>\n",
       "      <td>0.359106</td>\n",
       "      <td>0.448556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Garb_01G000020</th>\n",
       "      <td>2930</td>\n",
       "      <td>1.019808</td>\n",
       "      <td>242.78668</td>\n",
       "      <td>169.25060</td>\n",
       "      <td>0.000918</td>\n",
       "      <td>1_166471</td>\n",
       "      <td>148289</td>\n",
       "      <td>98</td>\n",
       "      <td>188</td>\n",
       "      <td>0.520619</td>\n",
       "      <td>0.000414</td>\n",
       "      <td>0.257425</td>\n",
       "      <td>0.071629</td>\n",
       "      <td>0.186813</td>\n",
       "      <td>0.192107</td>\n",
       "      <td>0.391936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Garb_01G000030</th>\n",
       "      <td>2984</td>\n",
       "      <td>1.009826</td>\n",
       "      <td>192.33210</td>\n",
       "      <td>159.18356</td>\n",
       "      <td>0.001345</td>\n",
       "      <td>1_195640</td>\n",
       "      <td>171061</td>\n",
       "      <td>12</td>\n",
       "      <td>22</td>\n",
       "      <td>0.056701</td>\n",
       "      <td>0.000428</td>\n",
       "      <td>-0.562865</td>\n",
       "      <td>0.157021</td>\n",
       "      <td>0.232767</td>\n",
       "      <td>0.223964</td>\n",
       "      <td>0.391936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Garb_01G000040</th>\n",
       "      <td>2989</td>\n",
       "      <td>1.082678</td>\n",
       "      <td>160.75177</td>\n",
       "      <td>150.88371</td>\n",
       "      <td>0.000242</td>\n",
       "      <td>1_491189</td>\n",
       "      <td>458872</td>\n",
       "      <td>22</td>\n",
       "      <td>42</td>\n",
       "      <td>0.116022</td>\n",
       "      <td>0.000034</td>\n",
       "      <td>-0.486441</td>\n",
       "      <td>0.114659</td>\n",
       "      <td>0.023976</td>\n",
       "      <td>0.028057</td>\n",
       "      <td>0.196402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Garb_01G000050</th>\n",
       "      <td>3021</td>\n",
       "      <td>1.043395</td>\n",
       "      <td>225.35234</td>\n",
       "      <td>164.27260</td>\n",
       "      <td>0.003105</td>\n",
       "      <td>1_455015</td>\n",
       "      <td>413933</td>\n",
       "      <td>102</td>\n",
       "      <td>197</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.001386</td>\n",
       "      <td>-0.234446</td>\n",
       "      <td>0.072248</td>\n",
       "      <td>0.510490</td>\n",
       "      <td>0.483138</td>\n",
       "      <td>0.483138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Garb_01G000060</th>\n",
       "      <td>3044</td>\n",
       "      <td>0.975400</td>\n",
       "      <td>154.70288</td>\n",
       "      <td>154.78766</td>\n",
       "      <td>0.001173</td>\n",
       "      <td>1_3045</td>\n",
       "      <td>-42649</td>\n",
       "      <td>53</td>\n",
       "      <td>103</td>\n",
       "      <td>0.737245</td>\n",
       "      <td>0.000300</td>\n",
       "      <td>0.294419</td>\n",
       "      <td>0.079939</td>\n",
       "      <td>0.178821</td>\n",
       "      <td>0.175072</td>\n",
       "      <td>0.391936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Garb_01G000070</th>\n",
       "      <td>3101</td>\n",
       "      <td>1.071897</td>\n",
       "      <td>206.42955</td>\n",
       "      <td>160.35652</td>\n",
       "      <td>0.002625</td>\n",
       "      <td>1_458889</td>\n",
       "      <td>404529</td>\n",
       "      <td>18</td>\n",
       "      <td>35</td>\n",
       "      <td>0.094595</td>\n",
       "      <td>0.000992</td>\n",
       "      <td>0.420298</td>\n",
       "      <td>0.125675</td>\n",
       "      <td>0.394605</td>\n",
       "      <td>0.384476</td>\n",
       "      <td>0.448556</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                num_var  beta_shape1  beta_shape2    true_df  pval_true_df  \\\n",
       "phenotype_id                                                                 \n",
       "Garb_01G000010     2910     1.060317    167.80818  154.71806      0.002920   \n",
       "Garb_01G000020     2930     1.019808    242.78668  169.25060      0.000918   \n",
       "Garb_01G000030     2984     1.009826    192.33210  159.18356      0.001345   \n",
       "Garb_01G000040     2989     1.082678    160.75177  150.88371      0.000242   \n",
       "Garb_01G000050     3021     1.043395    225.35234  164.27260      0.003105   \n",
       "Garb_01G000060     3044     0.975400    154.70288  154.78766      0.001173   \n",
       "Garb_01G000070     3101     1.071897    206.42955  160.35652      0.002625   \n",
       "\n",
       "               variant_id  tss_distance  ma_samples  ma_count        af  \\\n",
       "phenotype_id                                                              \n",
       "Garb_01G000010   1_360122        347814          63       117  0.297927   \n",
       "Garb_01G000020   1_166471        148289          98       188  0.520619   \n",
       "Garb_01G000030   1_195640        171061          12        22  0.056701   \n",
       "Garb_01G000040   1_491189        458872          22        42  0.116022   \n",
       "Garb_01G000050   1_455015        413933         102       197  0.500000   \n",
       "Garb_01G000060     1_3045        -42649          53       103  0.737245   \n",
       "Garb_01G000070   1_458889        404529          18        35  0.094595   \n",
       "\n",
       "                pval_nominal     slope  slope_se  pval_perm  pval_beta  \\\n",
       "phenotype_id                                                             \n",
       "Garb_01G000010      0.000912  0.271290  0.080528   0.373626   0.359106   \n",
       "Garb_01G000020      0.000414  0.257425  0.071629   0.186813   0.192107   \n",
       "Garb_01G000030      0.000428 -0.562865  0.157021   0.232767   0.223964   \n",
       "Garb_01G000040      0.000034 -0.486441  0.114659   0.023976   0.028057   \n",
       "Garb_01G000050      0.001386 -0.234446  0.072248   0.510490   0.483138   \n",
       "Garb_01G000060      0.000300  0.294419  0.079939   0.178821   0.175072   \n",
       "Garb_01G000070      0.000992  0.420298  0.125675   0.394605   0.384476   \n",
       "\n",
       "                    qval  \n",
       "phenotype_id              \n",
       "Garb_01G000010  0.448556  \n",
       "Garb_01G000020  0.391936  \n",
       "Garb_01G000030  0.391936  \n",
       "Garb_01G000040  0.196402  \n",
       "Garb_01G000050  0.483138  \n",
       "Garb_01G000060  0.391936  \n",
       "Garb_01G000070  0.448556  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cis_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trans-QTL mapping\n",
      "  * 197 samples\n",
      "  * 7 phenotypes\n",
      "  * 3 covariates\n",
      "  * 2143543 variants\n",
      "    processing batch 108/108\n",
      "    elapsed time: 0.47 min\n",
      "  * 1410773 variants passed MAF >= 0.05 filtering\n",
      "done.\n"
     ]
    }
   ],
   "source": [
    "#TODO 进行trans-eQTL分析\n",
    "#! 保留所有显著的位点\n",
    "trans_df = trans.map_trans(genotype_df, phenotype_df, covariates_df,\n",
    "                           return_sparse=True, pval_threshold=1, maf_threshold=0.05,\n",
    "                           batch_size=20000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "trans_df=trans_df.loc[trans_df['phenotype_id']=='Garb_01G000040']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "trans_df.to_csv(\"Garb_01G000040_trans_MAF.5.txt\",header=True,index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "66aa2addb38c89a0d43f63cbdb966cb2dba790ca2bb4fa0a3d79065325ea50af"
  },
  "kernelspec": {
   "display_name": "Python 3.9.13 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
